File: backend/app/schemas/voice.py
"""Request/response schemas for the voice query feature.

These are NEW schemas (see CLAUDE.md's frozen-contract rule) — they do not
modify any existing schema in this package.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.common import Language


class VoiceQueryRequest(BaseModel):
    run_id: str
    query: str = Field(..., min_length=1, max_length=300)
    language: Language = Language.en
    current_hour_index: int | None = None

    @field_validator("query")
    @classmethod
    def validate_query(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("query cannot be empty or whitespace only")
        return stripped


class VoiceQueryResponse(BaseModel):
    run_id: str
    language: Language
    intent: str
    answer: str
    facts_used: dict
    hour_index: int | None = None
    confidence: Literal["high", "low"]


class SpeakRequest(BaseModel):
    """POST /api/voice/speak — reads text aloud via Sarvam AI Bulbul TTS.

    The response is NOT JSON: it's a raw audio body (see
    app.services.tts.sarvam) with a matching Content-Type, so there is no
    SpeakResponse schema to mirror on the frontend.
    """

    text: str = Field(..., min_length=1, max_length=2500)
    language: Language = Language.en

    @field_validator("text")
    @classmethod
    def validate_text(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("text cannot be empty or whitespace only")
        return stripped

File: frontend/src/types/voice.ts
// Mirrors backend/app/schemas/voice.py 1:1 (see CLAUDE.md API Contract).

import type { Language } from './api'

export interface VoiceQueryRequest {
  run_id: string
  query: string
  language: Language
  current_hour_index?: number | null
}

export interface VoiceQueryResponse {
  run_id: string
  language: Language
  intent: string
  answer: string
  facts_used: Record<string, unknown>
  hour_index: number | null
  confidence: 'high' | 'low'
}

// The SSE payloads sent by POST /api/voice/stream — not Pydantic models
// (there is no response body to mirror), but kept here so the frontend has
// one typed source for every event shape on that stream.
export interface VoiceIntentEvent {
  intent: string
  confidence: 'high' | 'low'
  hour_index: number | null
}

export interface VoiceTokenEvent {
  text: string
}

export interface VoiceDoneEvent {
  answer: string
  facts_used: Record<string, unknown>
  validated: boolean
}

export interface VoiceErrorEvent {
  message: string
}

// POST /api/voice/speak — Sarvam AI Bulbul TTS. The response is a raw audio
// body (not JSON), so there is no matching response interface to mirror.
export interface SpeakRequest {
  text: string
  language: Language
}
